### PR TITLE
[WIP] Voila as an ExtensionApp

### PR DESCRIPTION
**Pull request by Zsailer**
I went ahead and started refactoring Voila to use the [`ExtensionApp`](https://github.com/jupyter/jupyter_server/pull/48) in jupyter_server. It greatly simplifies the main application. 

Using this API, Voila can be initialized, configured and launched from the command line, and you can generate configuration files using the `--generate-config` flag.

It can also be loaded as a Jupyter server extension, allowing you to serve Voila side-by-side with other client applications running on one server. 

One thing to note is that all static files are now served behind the url namespace `static/voila/` (this is automatically set up by `ExtensionApp`).

This should be working. Let me know if you have issues with testing it out.  

@SylvainCorlay @maartenbreddels

### CODE REVIEW

**Comment by Zsailer** [comment]
To test this out,

1. Clone and install jupyter_server master (the latest release won't work).
2. Checkout this PR's branch
3. Install using `pip install .` (this updates the voila CLI and server extension entry points).

Start Voila by either 1) running directly:
```
jupyter voila
```
or loading as a server extension
```
jupyter server --ServerApp.default_url='/voila'`
```

**Comment by Zsailer** [comment]
Also, I've opened jupyter/jupyter_server#59 to simplify how extensions configuration gets passed to extension handlers. This should help simplify some of the changes here. 

Once that PR is merged in jupyter_server, I'll update this PR.

**Comment by timkpaine** [comment]
checking this now

**Comment by Zsailer** [comment]
Thanks, @timkpaine.

I think this will require a pretty significant refactor of Voila's tests. I'm happy to work on this if you can confirm that the `ExtensionApp` seems like a good solution for Voila.

**Comment by timkpaine** [comment]
@Zsailer I really would like to use my custom jupyterlab `LoginHandler` for Voila, rather than having to do any custom weirdness, so I think this makes sense to do. Im not a maintainer of voila though so its up for @SylvainCorlay to decide.

**Comment by Zsailer** [comment]
>I really would like to use my jupyterlab LoginHandler for Voila

Definitely—though, you may have to wait for JupyterLab to adopt `jupyter_server` as it's core server (unless you're already using Lab without the classic notebook server).

**Comment by timkpaine** [comment]
I don't need full compatibility, just the bare minimum

File: setup.py [hunk 1]
```diff
@@ -273,7 +273,7 @@ def get_data_files():
     },
     'entry_points': {
         'console_scripts': [
-            'voila = voila.app:main'
+            'jupyter-voila = voila.app:main'
```
Comment (maartenbreddels) [inline setup.py]: Do we want this in addition to `voila` for backward compatibility?

Comment (SylvainCorlay) [reply setup.py]: We will probably want to have the standalone mode supported in `jupyter_server` before we make  `voila`  be the same thing as `jupyter-voila`.

Comment (Zsailer) [reply setup.py]: :+1: I'll probably work on the standalone mode later this week.

**Comment by Zsailer** [comment]
@maartenbreddels Great! Glad you like it!

I just got back from holiday, so I'm catching up on emails+Github threads. I'm hoping to rebase this PR later this week. It should work with the latest jupyter_server, but I'll verify and ping you again soon.

File: setup.py (hunk 1 above)
Comment (Zsailer) [reply setup.py]: added "standalone" mode in jupyter/jupyter_server#70

**Comment by Zsailer** [comment]
Ok, once jupyter/jupyter_server#70 is merged, this should be working (with standalone mode).

File: voila/app.py [hunk 2]
```diff
@@ -112,51 +104,48 @@ class Voila(Application):
...
+        help=(
+            'template name to be used by voila.'
         )
     )
+    resources = Dict(
+        allow_none=True,
+        help="""
+        extra resources used by templates;
+        example use with --template=reveal
+        --VoilaConfiguration.resources="{'reveal': {'transition': 'fade', 'scroll': True}}"
```
Comment (maartenbreddels) [inline voila/app.py]:
```suggestion
        --Voila.resources="{'reveal': {'transition': 'fade', 'scroll': True}}"
```

**Comment by Zsailer** [comment]
The tests need some refactoring after these significant changes. I started working on that last week but didn't quite finish. I'll try to update the tests over the next day or two.

**Comment by Zsailer** [comment]
> I also noticed that if I run `jupyter-voila notebooks/basics.ipynb`,

Good catch! Looking into this issue now.

**Comment by vidartf** [comment]
I believe this would also be relevant? https://github.com/jupyter/jupyter_server/issues/80

**Comment by timkpaine** [comment]
@Zsailer anything I can do to help?

**Comment by Zsailer** [comment]
@timkpaine 

I'm waiting for jupyter/jupyter_server#76 to land. In the meantime, I've rebased here and starting working on tests. 

There's quite a bit of work to do on the tests. This PR rips out the server portion of the app. Many of the tests try to change server traits—I think that's out of scope now. Those tests should live in jupyter_server, no?

File: voila/static_file_handler.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ def initialize(self, paths, default_filename=None):
     def get_absolute_path(self, root, path):
         # find the first absolute path that exists
         self.root = self.roots[0]
-        for root in self.roots:
+        for root in self.roots:git
```
Comment (vidartf) [inline voila/static_file_handler.py]: I'm guessing this shouldn't be here 😉

Comment (timkpaine) [reply voila/static_file_handler.py]: hahahah I do this all the time

Comment (Zsailer) [reply voila/static_file_handler.py]: 🤦‍♂

Comment (Zsailer) [reply voila/static_file_handler.py]: 😆

**Comment by Zsailer** [comment]
@timkpaine and @vidartf 

I'm working on this all day today. Hoping to have the tests passing by the end of the day (~5pm PDT).

**Comment by timkpaine** [comment]
@Zsailer feel free to take whatever from https://github.com/vidartf/phoila

**Comment by Zsailer** [comment]
Almost finished! 

I have a few more tests to get passing, but I should be finished early tomorrow. Most tests required the addition of auth tokens to their http requests—a new feature from using jupyter_server under the hood. 

One the last few tests are finished, I'll rebase and this should be ready for detail review.

File: voila/static_file_handler.py (hunk 3 above)
Comment (vidartf) [reply voila/static_file_handler.py]: Just be glad it wasn't a password (right??) 😄

**Comment by Zsailer** [comment]
I've got two more tests to get working (I got pulled elsewhere today). I'll continue working on this tomorrow. So close!

- [ ] `tests/app/execute_test.py`
- [ ] `tests/app/nbextensions_test.py`

**Comment by vidartf** [comment]
> Yes, this PR can make use of jupyter/jupyter_server#80. Question—you only want to drop contents when running voila in "standalone" mode, right?
> 
> In contrast, when one runs voila side-by-side with other frontends, like jupyterlab, they'd want those services exposed. So it depends on how the extension is started, right?
> 
> We can certainly change the default_services of the server in voila's standalone mode. That would be a small modification to its launch_instance method.

I think you got the gist of it (and it is how voila currently works):
- If used as an extension, leave existing handlers as they are (but add new ones)
- If used as an app, ensure only a limited subset of handlers are available.

**Comment by vidartf** [comment]
Note, CI is failing on the lint: `flake8 voila tests setup.py`

**Comment by vidartf** [comment]
PS: Note I already have one inline review comment above (about `notebook_path` vs `file_to_run`).

**Comment by Zsailer** [comment]
Fixed flake8.

**Comment by Zsailer** [comment]
Tests are failing because jupyter_server needs a new release.

File: voila/app.py [hunk 4]
```diff
@@ -315,216 +261,82 @@ def _default_root_dir(self):
...
+        super(Voila, self).parse_command_line(argv)
+        # Catch first argument as notebook path.
+        if self.extra_args:
+            notebook_path = self.extra_args[0]
+            f = os.path.abspath(notebook_path)
+            self.argv.remove(notebook_path)
+            if not os.path.exists(f):
+                self.log.critical(_("No such file or directory: %s"), f)
+                self.exit(1)
+            self.notebook_path = notebook_path
```
[the remark this answers is not in the document]
Comment (Zsailer) [reply voila/app.py]: Yes, it does. I meant to change this to using ServerApp's file_to_run. I've updated by making the default value for `notebook_path` to be pulled from ServerApp's file_to_run.

**Comment by echarles** [comment]
Built this branch and launched `jupyter-voila` (to invoke the entry_point) => `Exception: The extension must be given a `name`.

Merged latest commit (fixed minor conflict) => `voila` entry_point is not created.

**Comment by maartenbreddels** [comment]
@Zsailer would you mind if I do a rebase (and force push), and do some iterations on this branch?

**Comment by Zsailer** [comment]
@maartenbreddels not at all. Go for it! Sorry I've been behind—trying to carve time to work on this.

**Comment by Zsailer** [comment]
@maartenbreddels this looks great. I'm waiting to cut a new release of jupyter_server, then I'll update this PR with some recent changes there. Then, I think this is ready to go!

**Comment by Zsailer** [comment]
Closing in favor of #492. Thanks, @maartenbreddels!

**Pull request closed without merging**